### PR TITLE
[JENKINS-72067] High memory usage from `XStream2.AssociatedConverterImpl`

### PR DESCRIPTION
**Pull request by basil**
See [JENKINS-72067](https://issues.jenkins.io/browse/JENKINS-72067). Partial revert of #7976, restoring previous memory usage levels.

### Testing done

Reproduced the problem as described in Jira; could no longer reproduce after this PR.

### Proposed changelog entries

- Reduce high memory usage from `XStream2.AssociatedConverterImpl` (regression in 2.405).

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Review by jglick — APPROVED** [review]
IIUC keeping `classCache` and the `ClassLoader.getResource` optimization, but reverting (non-`static`) `cache` with some modifications. Seems fine to me—whatever works. It is not immediately clear to me from the bug report what concretely was causing the leak. Perhaps `ClassValue` is not designed for use in a non-`static` context.

(As noted in #7976 the use of the strongly held keys would prevent `Class`es with associated converters from being collected, though this is unlikely to matter in practice as Jenkins does not support dynamic plugin unloading, and classes loaded from e.g. Groovy scripts should not be participating in XStream serialization.)

**Review by NotMyFault — APPROVED** [review]

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by MarkEWaite — APPROVED** [review]
I tested the incremental build from this pull request with the setup as described in [JENKINS-72067](https://issues.jenkins.io/browse/JENKINS-72067).  With 1001 runs of a Pipeline job using git scm and 1001 runs of a Pipeline job using file system SCM the memory growth reported by the pre-release was significantly less than the growth in the original report.  Instead of over 100 MB memory growth in ConcurrentHashMaps, I saw only 4 MB of memory growth in ConcurrentHashMaps.  That growth is consistent with what was observed in 2.404.

This is ready to go.  Thanks to all!

**Pull request merged**
